File: java/src/main/java/com/fixcraft/basefwx/Format.java
package com.fixcraft.basefwx;

import java.util.ArrayList;
import java.util.List;

public final class Format {
    private Format() {}

    // Pick a sane cap for your app (or make it configurable).
    // For encryption formats, 64 MiB for metadata blobs is usually plenty.
    public static final int MAX_TOTAL = 64 * 1024 * 1024;

    public static byte[] packLengthPrefixed(List<byte[]> parts) {
        if (parts == null) throw new IllegalArgumentException("parts == null");

        long total = 4L * parts.size();
        for (byte[] part : parts) {
            if (part == null) throw new IllegalArgumentException("null part");
            total += part.length;
            if (total > MAX_TOTAL) {
                throw new IllegalArgumentException("length-prefixed blob too large");
            }
        }

        byte[] out = new byte[(int) total];
        int offset = 0;

        for (byte[] part : parts) {
            int len = part.length;
            out[offset]     = (byte) (len >>> 24);
            out[offset + 1] = (byte) (len >>> 16);
            out[offset + 2] = (byte) (len >>> 8);
            out[offset + 3] = (byte) (len);
            offset += 4;

            if (len != 0) {
                System.arraycopy(part, 0, out, offset, len);
                offset += len;
            }
        }
        return out;
    }

    public static List<byte[]> unpackLengthPrefixed(byte[] data, int count) {
        if (data == null) throw new IllegalArgumentException("data == null");
        if (count < 0) throw new IllegalArgumentException("count < 0");
        if (data.length > MAX_TOTAL) throw new IllegalArgumentException("blob too large");

        List<byte[]> parts = new ArrayList<>(count);
        int offset = 0;

        for (int i = 0; i < count; i++) {
            if (offset + 4 > data.length) {
                throw new IllegalArgumentException("Malformed length-prefixed blob (missing length)");
            }

            int len = ((data[offset] & 0xFF) << 24)
                    | ((data[offset + 1] & 0xFF) << 16)
                    | ((data[offset + 2] & 0xFF) << 8)
                    | (data[offset + 3] & 0xFF);
            offset += 4;

            // Critical: reject negative (and also crazy large).
            if (len < 0) {
                throw new IllegalArgumentException("Malformed length-prefixed blob (negative length)");
            }
            if (len > MAX_TOTAL) {
                throw new IllegalArgumentException("Malformed length-prefixed blob (part too large)");
            }
            if (offset > data.length - len) { // avoids offset+len overflow
                throw new IllegalArgumentException("Malformed length-prefixed blob (truncated part)");
            }

            byte[] part = new byte[len];
            if (len != 0) {
                System.arraycopy(data, offset, part, 0, len);
                offset += len;
            }
            parts.add(part);
        }

        if (offset != data.length) {
            throw new IllegalArgumentException("Malformed length-prefixed blob (extra bytes)");
        }
        return parts;
    }
}

File: cpp/src/formats/format.cpp
/*
 * BaseFWX - Cryptography Engine
 * Copyright (C) 2020-2026  FixCraft Inc.
 * Licensed under the GNU Lesser General Public License v3.0 or later.
 */

#include "basefwx/format.hpp"

#include "basefwx/base64.hpp"
#include "basefwx/constants.hpp"

#include <limits>
#include <stdexcept>

namespace basefwx::format {

namespace {

std::uint32_t ReadU32BE(const Bytes& data, std::size_t offset) {
    if (offset + 4 > data.size()) {
        throw std::runtime_error("Malformed length-prefixed blob (missing length)");
    }
    return (static_cast<std::uint32_t>(data[offset]) << 24)
           | (static_cast<std::uint32_t>(data[offset + 1]) << 16)
           | (static_cast<std::uint32_t>(data[offset + 2]) << 8)
           | static_cast<std::uint32_t>(data[offset + 3]);
}

}  // namespace

Bytes PackLengthPrefixed(const std::vector<Bytes>& parts) {
    constexpr std::size_t max =
        static_cast<std::size_t>(basefwx::constants::kLengthPrefixedMax);
    if (parts.size() > max / 4u) {
        throw std::runtime_error(
            "Length-prefixed blob exceeds 64 MiB total cap");
    }
    std::size_t total = 4u * parts.size();
    for (const auto& part : parts) {
        if (part.size() > max
            || part.size() > std::numeric_limits<std::uint32_t>::max()) {
            throw std::runtime_error(
                "Length-prefixed part exceeds 64 MiB cap");
        }
        if (part.size() > max - total) {
            throw std::runtime_error(
                "Length-prefixed blob exceeds 64 MiB total cap");
        }
        total += part.size();
    }
    Bytes out(total);
    std::size_t offset = 0;
    for (const auto& part : parts) {
        std::uint32_t len = static_cast<std::uint32_t>(part.size());
        out[offset] = static_cast<std::uint8_t>((len >> 24) & 0xFF);
        out[offset + 1] = static_cast<std::uint8_t>((len >> 16) & 0xFF);
        out[offset + 2] = static_cast<std::uint8_t>((len >> 8) & 0xFF);
        out[offset + 3] = static_cast<std::uint8_t>(len & 0xFF);
        offset += 4;
        if (!part.empty()) {
            std::copy(part.begin(), part.end(), out.begin() + offset);
            offset += part.size();
        }
    }
    return out;
}

std::vector<Bytes> UnpackLengthPrefixed(const Bytes& data, std::size_t count) {
    // Match the 64 MiB total cap that Format.java has had since 3.4.x.
    // Without this, a malicious blob declaring a single 4 GiB part survives
    // until the data.size() bounds check fires — long enough for any
    // upstream code that pre-sizes a buffer from the length field to OOM.
    constexpr std::size_t kMaxPartLen =
        static_cast<std::size_t>(basefwx::constants::kLengthPrefixedMax);
    constexpr std::size_t kMaxTotalLen =
        static_cast<std::size_t>(basefwx::constants::kLengthPrefixedMax);
    if (data.size() > kMaxTotalLen) {
        throw std::runtime_error(
            "Length-prefixed blob exceeds 64 MiB total cap");
    }
    std::vector<Bytes> parts;
    parts.reserve(count);
    std::size_t offset = 0;
    std::size_t total_consumed = 0;
    for (std::size_t i = 0; i < count; ++i) {
        std::uint32_t len = ReadU32BE(data, offset);
        offset += 4;
        if (len > kMaxPartLen) {
            throw std::runtime_error("Length-prefixed part exceeds 64 MiB cap");
        }
        total_consumed += len;
        if (total_consumed > kMaxTotalLen) {
            throw std::runtime_error("Length-prefixed blob exceeds 64 MiB total cap");
        }
        if (offset + len > data.size()) {
            throw std::runtime_error("Malformed length-prefixed blob (truncated part)");
        }
        Bytes part;
        if (len > 0) {
            part.insert(part.end(), data.begin() + offset, data.begin() + offset + len);
        }
        parts.push_back(std::move(part));
        offset += len;
    }
    if (offset != data.size()) {
        throw std::runtime_error("Malformed length-prefixed blob (extra bytes)");
    }
    return parts;
}

PayloadParts SplitPayload(const Bytes& payload) {
    if (payload.size() < 4) {
        throw std::runtime_error("Payload too short");
    }
    std::uint32_t meta_len = ReadU32BE(payload, 0);
    if (meta_len > basefwx::constants::kMetadataMax) {
        throw std::runtime_error("Payload metadata exceeds 1 MiB cap");
    }
    std::size_t meta_end = 4 + meta_len;
    if (meta_end > payload.size()) {
        throw std::runtime_error("Malformed payload metadata header");
    }
    PayloadParts parts;
    if (meta_len > 0) {
        parts.metadata_blob.assign(payload.begin() + 4, payload.begin() + meta_end);
    }
    parts.ciphertext.insert(parts.ciphertext.end(), payload.begin() + meta_end, payload.end());
    return parts;
}

std::optional<MetadataPreview> TryDecodeMetadata(const Bytes& payload) {
    if (payload.size() < 4) {
        return std::nullopt;
    }
    std::uint32_t meta_len = ReadU32BE(payload, 0);
    if (meta_len > basefwx::constants::kMetadataMax) {
        return std::nullopt;
    }
    std::size_t meta_end = 4 + meta_len;
    if (meta_end > payload.size()) {
        return std::nullopt;
    }
    MetadataPreview preview;
    preview.metadata_len = meta_len;
    if (meta_len == 0) {
        return preview;
    }
    std::string blob(payload.begin() + 4, payload.begin() + meta_end);
    if (!basefwx::base64::IsLikelyBase64(blob)) {
        return std::nullopt;
    }
    bool ok = false;
    std::vector<std::uint8_t> decoded = basefwx::base64::Decode(blob, &ok);
    if (!ok) {
        return std::nullopt;
    }
    preview.metadata_base64 = blob;
    preview.metadata_json.assign(decoded.begin(), decoded.end());
    return preview;
}

}  // namespace basefwx::format
